### PR TITLE
Add conditional logic for dataset triggering

### PR DESCRIPTION
**Pull request by dstandish**
Add conditional logic for dataset-triggered dags.

This means we can schedule based on `dataset1 OR dataset1`.

This PR only implements the underlying classes, `DatasetAny` and `DatasetAll`.  In a followup PR we will add more convenient syntax for this, specifically the `|` and `&` symbols, e.g. `(dataset1 | dataset2) & dataset3`.

### CODE REVIEW

**Comment by dstandish** [comment]
@sunank200 

1. fix test on main PR
2. look at the tests on the PR and determine which should be added or removed
3. performance
	- why 10 minutes in main?
	- why slower with this PR?
	- what's going on?

**Comment by sunank200** [comment]
> @sunank200
> 
> 1. fix test on main PR
> 2. look at the tests on the PR and determine which should be added or removed
> 3. performance
>    
>    * why 10 minutes in main?
>    * why slower with this PR?
>    * what's going on?

Documentation changes are done [here](https://github.com/apache/airflow/pull/37101)

File: airflow/jobs/scheduler_job_runner.py [hunk 1]
```diff
@@ -1247,59 +1240,68 @@ def _create_dag_runs_dataset_triggered(
...
+                dataset_event_filters.append(DatasetEvent.timestamp > prev_exec_date)
+
+            dataset_events = session.scalars(
+                select(DatasetEvent)
+                .join(
+                    DagScheduleDatasetReference,
+                    DatasetEvent.dataset_id == DagScheduleDatasetReference.dataset_id,
                 )
+                .join(DatasetEvent.source_dag_run)
+                .where(*dataset_event_filters)
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: I kind of wonder if we can eliminate the extra `prev_exec_date` query, but that can probably be a separate improvement later.

File: tests/serialization/test_dag_serialization.py [hunk 2]
```diff
@@ -550,6 +549,7 @@ def validate_deserialized_dag(self, serialized_dag, dag):
             "_task_group",
             "params",
             "_processor_dags_folder",
+            "dataset_triggers",
```
Comment (uranusjr) [inline tests/serialization/test_dag_serialization.py:555]: Is it not possible to check this?

File: airflow/jobs/scheduler_job_runner.py (hunk 1 above)
Comment (dstandish) [reply airflow/jobs/scheduler_job_runner.py]: yeah i'll leave this one alone for now.

File: tests/serialization/test_dag_serialization.py (hunk 2 above)
Comment (dstandish) [reply tests/serialization/test_dag_serialization.py:555]: possible; enabled

File: airflow/models/dataset.py [hunk 3]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
...
+
+    :meta private:
+    """
+
+    agg_func: Callable
+
+    def __init__(self, *objects):
+        self.objects = objects
+
+    def evaluate(self, statuses: dict[str, bool]):
```
Comment (jedcunningham) [inline airflow/models/dataset.py]:
```suggestion
    def evaluate(self, statuses: dict[str, bool]) -> bool:
```

File: airflow/models/dataset.py [hunk 4]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
...
+        return self.agg_func(self.eval_one(x, statuses) for x in self.objects)
+
+    def eval_one(self, obj: Dataset | DatasetAny | DatasetAll, statuses):
```
Comment (jedcunningham) [inline airflow/models/dataset.py]:
```suggestion
    def eval_one(self, obj: Dataset | DatasetAny | DatasetAll, statuses) -> bool:
```

File: airflow/models/dataset.py [hunk 5]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
...
+class DatasetBooleanCondition:
+    """
+    Base class for boolean logic for dataset triggers.
+
+    :meta private:
+    """
+
+    agg_func: Callable
+
+    def __init__(self, *objects):
```
Comment (jedcunningham) [inline airflow/models/dataset.py]:
```suggestion
    def __init__(self, *objects) -> None:
```

File: airflow/timetables/datasets.py [hunk 6]
```diff
@@ -52,24 +55,20 @@ def deserialize(cls, data: dict[str, typing.Any]) -> Timetable:
         from airflow.serialization.serialized_objects import decode_timetable
 
         return cls(
-            timetable=decode_timetable(data["timetable"]), datasets=[Dataset(**d) for d in data["datasets"]]
+            timetable=decode_timetable(data["timetable"]),
+            datasets=[],  # don't need the datasets after deserialization
```
Comment (jedcunningham) [inline airflow/timetables/datasets.py]: Why don't we need them?

File: airflow/example_dags/example_datasets.py [hunk 7]
```diff
@@ -50,6 +50,7 @@
 dag1_dataset = Dataset("s3://dag1/output_1.txt", extra={"hi": "bye"})
 # [END dataset_def]
 dag2_dataset = Dataset("s3://dag2/output_1.txt", extra={"hi": "bye"})
+dag3_dataset = Dataset("s3://dag3/output_3.txt", extra={"hi": "bye"})
```
Comment (bbovenzi) [inline airflow/example_dags/example_datasets.py]: Are we supposed to also use this dataset as an any/all rule in one of these dags?

File: tests/cli/commands/test_dag_command.py
Comment (jedcunningham) [inline tests/cli/commands/test_dag_command.py:1]: How are these changes related?

File: tests/datasets/test_dataset.py [hunk 8]
```diff
@@ -40,17 +46,224 @@ def test_invalid_uris(uri):
...
+
+    for (serialized_dag,) in serialized_dags:
+        dag = SerializedDAG.deserialize(serialized_dag.data)
+        for dataset_uri, status in dag_statuses[dag.dag_id].items():
+            assert dag.dataset_triggers.evaluate({dataset_uri: status}), "DAG trigger evaluation failed"
+
+
+@pytest.mark.db_test
+@pytest.mark.usefixtures("create_test_datasets")
+def test_additional_dag_with_no_triggers(dag_maker):
```
Comment (jedcunningham) [inline tests/datasets/test_dataset.py]: This test feels a little odd without some additional assertions?

File: tests/datasets/test_dataset.py [hunk 9]
```diff
@@ -40,17 +46,224 @@ def test_invalid_uris(uri):
...
+    # Create an additional DAG to ensure it's not affected by dataset triggers
+    with dag_maker(dag_id="dag2"):
+        EmptyOperator(task_id="hello2")
+
+
+@pytest.fixture
+def setup_datasets_and_models(session):
```
Comment (jedcunningham) [inline tests/datasets/test_dataset.py]: Not sure this needs to be a fixture, it's only used once and isn't that big.

File: tests/datasets/test_dataset.py [hunk 10]
```diff
@@ -40,17 +46,224 @@ def test_invalid_uris(uri):
...
+
+    serialized_dag_dict = SerializedDAG.to_dict(dag)["dag"]
+    assert "dataset_triggers" in serialized_dag_dict, "Serialized DAG should contain 'dataset_triggers'"
+    assert isinstance(
+        serialized_dag_dict["dataset_triggers"], dict
+    ), "Serialized 'dataset_triggers' should be a dict"
+
+
+@pytest.fixture(autouse=True)
+def clear_datasets():
```
Comment (jedcunningham) [inline tests/datasets/test_dataset.py]: nit: move this to the top.

File: tests/datasets/test_dataset.py [hunk 11]
```diff
@@ -40,17 +46,224 @@ def test_invalid_uris(uri):
...
+    from tests.test_utils.db import clear_db_datasets
+
+    clear_db_datasets()
```
Comment (jedcunningham) [inline tests/datasets/test_dataset.py]:
```suggestion
    clear_db_datasets()
    yield
    clear_db_datasets()
```

Should we also clear when we finish up?

File: tests/datasets/test_dataset.py [hunk 12]
```diff
@@ -18,13 +18,19 @@
...
+from sqlalchemy.sql import select
 
 from airflow.datasets import Dataset
+from airflow.models.dataset import DatasetAll, DatasetAny, DatasetDagRunQueue, DatasetModel
+from airflow.models.serialized_dag import SerializedDagModel
 from airflow.operators.empty import EmptyOperator
+from airflow.serialization.serialized_objects import BaseSerialization, SerializedDAG
 
 
+@pytest.mark.db_test
```
Comment (jedcunningham) [inline tests/datasets/test_dataset.py]: Why do all of these need to db tests now?

File: airflow/models/dataset.py [hunk 13]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
         ]:
             args.append(f"{attr}={getattr(self, attr)!r}")
         return f"{self.__class__.__name__}({', '.join(args)})"
+
+
+class DatasetBooleanCondition:
```
Comment (Lee-W) [inline airflow/models/dataset.py:342]: Looks like this is not something we're going to use directly. Should we make it an abstract class?

File: airflow/models/dataset.py [hunk 14]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
...
+    """
+    Base class for boolean logic for dataset triggers.
+
+    :meta private:
+    """
+
+    agg_func: Callable
```
Comment (Lee-W) [inline airflow/models/dataset.py]:
```suggestion
    agg_func: Callable[[Iterable], bool]
```

File: airflow/models/dataset.py [hunk 15]
```diff
@@ -336,3 +337,49 @@ def __repr__(self) -> str:
...
+
+    def evaluate(self, statuses: dict[str, bool]):
+        return self.agg_func(self.eval_one(x, statuses) for x in self.objects)
+
+    def eval_one(self, obj: Dataset | DatasetAny | DatasetAll, statuses):
+        if isinstance(obj, Dataset):
+            return statuses.get(obj.uri, False)
+        return obj.evaluate(statuses=statuses)
+
+    def all_datasets(self) -> dict[str, Dataset]:
```
Comment (Lee-W) [inline airflow/models/dataset.py:362]: Should it be a property or should we name it as something like `get_all_datasets`?

File: airflow/timetables/datasets.py (hunk 6 above)
Comment (Lee-W) [reply airflow/timetables/datasets.py]: same question here

**Comment by bbovenzi** [comment]
Do we plan to update `object/next_run_datasets` so the UI can show all of this logic?

Same for `/next_run_datasets_summary`, We should make sure the ready+total counts are still accurate. It might need to change from total to a `min` and `max`. so we can say something like `1 of 2-3 datasets updated`

File: airflow/timetables/datasets.py (hunk 6 above)
Comment (dstandish) [reply airflow/timetables/datasets.py]: they end up getting stored on a dataset_triggers attr of DAG. i can add a better comment.

File: tests/datasets/test_dataset.py (hunk 12 above)
Comment (dstandish) [reply tests/datasets/test_dataset.py]: I am not sure, @sunank200 added that in response to test failures i can take a look.

Comment (dstandish) [reply tests/datasets/test_dataset.py]: oh yeah it's cus an autouse fixture was added
```
@pytest.fixture(autouse=True)
def clear_datasets():
    from tests.test_utils.db import clear_db_datasets

    clear_db_datasets()
```

i'll remove that

Comment (sunank200) [reply tests/datasets/test_dataset.py]: That's true its for `clear_datasets()`

File: airflow/example_dags/example_datasets.py (hunk 7 above)
Comment (dstandish) [reply airflow/example_dags/example_datasets.py]: removed; this will be added in @sunank200's followup

File: airflow/models/dataset.py (hunk 15 above)
Comment (dstandish) [reply airflow/models/dataset.py:362]: it's a good suggestion.  this will be changed in followup refactor from TP, so i'll leave it for now to make rebasing easier if that's alright with you

File: tests/datasets/test_dataset.py (hunk 8 above)
Comment (sunank200) [reply tests/datasets/test_dataset.py]: I think the test_additional_dag_with_no_triggers function was to validate that a DAG runs correctly, without any dataset triggers and it is correctly configured and functions independently of dataset-related events.

File: airflow/models/dataset.py (hunk 13 above)
Comment (dstandish) [reply airflow/models/dataset.py:342]: This will be changed in TP's refactor so i'll leave it for now; in any case it's private so we can change freely

File: tests/datasets/test_dataset.py (hunk 12 above)
Comment (jedcunningham) [reply tests/datasets/test_dataset.py]: Gotcha. Let's be more targeted with what tests need to hit the db and which don't. db tests aren't as quick.

File: airflow/models/dataset.py (hunk 15 above)
Comment (dstandish) [reply airflow/models/dataset.py:362]: this will be changed in TP's followup refactor; i will leave for now to make rebase easier

File: tests/datasets/test_dataset.py (hunk 9 above)
Comment (dstandish) [reply tests/datasets/test_dataset.py]: yup, removing

File: airflow/timetables/datasets.py (hunk 6 above)
Comment (dstandish) [reply airflow/timetables/datasets.py]: added

File: tests/datasets/test_dataset.py (hunk 8 above)
Comment (dstandish) [reply tests/datasets/test_dataset.py]: but that's not running the dag.... it's just creating it

File: tests/datasets/test_dataset.py (hunk 12 above)
Comment (dstandish) [reply tests/datasets/test_dataset.py]: fixed it

File: tests/datasets/test_dataset.py (hunk 10 above)
Comment (dstandish) [reply tests/datasets/test_dataset.py]: moved

File: tests/cli/commands/test_dag_command.py
Comment (dstandish) [reply tests/cli/commands/test_dag_command.py:1]: the issue was the test was not good. they were testing that it works with custom table but instead of custom timetable they used a mock object. so when the test actually ran, it was actually using a different non-custom timetable. it was passing by accident.  our logic change re parsing of the "schedule" attr changed the behavior and revealed the issue, which we fixed by mocking less invasively.

**Comment by dstandish** [comment]
> Do we plan to update `object/next_run_datasets` so the UI can show all of this logic?
> 
> Same for `/next_run_datasets_summary`, We should make sure the ready+total counts are still accurate. It might need to change from total to a `min` and `max`. so we can say something like `1 of 2-3 datasets updated`

Created issue on our board @bbovenzi

**Review by jedcunningham — APPROVED** [review]
